Terminal.Gui/View/Layout/Dim.cs

```diff
@@ -0,0 +1,766 @@
+/// </remarks>
+/// <param name="percent">The percentage.</param>
+/// <param name="usePosition">
+///     If <see langword="true"/> the dimension is computed using the View's position (<see cref="View.X"/> or
+///     <see cref="View.Y"/>).
+///     If <see langword="false"/> the dimension is computed using the View's <see cref="View.ContentSize"/>.
+/// </param>
+public class DimPercent (float percent, bool usePosition = false) : Dim
+{
+    /// <inheritdoc/>
+    public override bool Equals (object other) { return other is DimPercent f && f.Percent == Percent && f.UsePosition == UsePosition; }
+
+    /// <inheritdoc/>
+    public override int GetHashCode () { return Percent.GetHashCode (); }
+
+    /// <summary>
+    ///     Gets the percentage.
+    /// </summary>
+    public new float Percent { get; } = percent;
+
+    /// <summary>
+    /// </summary>
+    /// <returns></returns>
+    public override string ToString () { return $"Percent({Percent},{UsePosition})"; }
+
+    /// <summary>
+    ///     Gets whether the dimension is computed using the View's position or ContentSize.
+    /// </summary>
+    public bool UsePosition { get; } = usePosition;
+
+    internal override int Anchor (int size) { return (int)(size * Percent); }
+
+    internal override int Calculate (int location, int superviewContentSize, View us, Dimension dimension)
+    {
+        return UsePosition ? Math.Max (Anchor (superviewContentSize - location), 0) : Anchor (superviewContentSize);
+    }
+}
+
+/// <summary>
+///     Represents a dimension that fills the dimension, leaving the specified margin.
+/// </summary>
+/// <remarks>
+///     This is a low-level API that is typically used internally by the layout system. Use the various static
+///     methods on the <see cref="Dim"/> class to create <see cref="Dim"/> objects instead.
+/// </remarks>
+/// <param name="margin">The margin to not fill.</param>
+public class DimFill (int margin) : Dim
+{
+    /// <inheritdoc/>
+    public override bool Equals (object other) { return other is DimFill fill && fill.Margin == Margin; }
+
+    /// <inheritdoc/>
+    public override int GetHashCode () { return Margin.GetHashCode (); }
+
+    /// <summary>
+    ///     Gets the margin to not fill.
+    /// </summary>
+    public int Margin { get; } = margin;
+
+    /// <inheritdoc/>
+    public override string ToString () { return $"Fill({Margin})"; }
+
+    internal override int Anchor (int size) { return size - Margin; }
+}
+
+/// <summary>
+///     Represents a function <see cref="Dim"/> object that computes the dimension by executing the provided function.
+/// </summary>
+/// <remarks>
+///     This is a low-level API that is typically used internally by the layout system. Use the various static
+///     methods on the <see cref="Dim"/> class to create <see cref="Dim"/> objects instead.
+/// </remarks>
+/// <param name="dim"></param>
+public class DimFunc (Func<int> dim) : Dim
```

Comment:
I keep meaning to bring up that this needs a more specific implementation, because it's problematic in a _lot_ of potential uses, when unrestricted like this.

Still a cool thing to have, but something we can address later on and that I've not had the opportunity to specifically think about, yet.

It's like...almost there, too. 🤔

Reply:
Please elaborate. I don't understand what "needs a more specific implementation" means.

Reply:
I will. It's way out of scope for this PR though, so I'll stick it in a new issue if it still looks like it matters later.

It's minor.